File: apps/admin-web/src/features/notifications/notification.service.ts
import { createClient } from '@/lib/supabase/client';
import { AppError, ErrorCode } from '@/lib/errors/app-error';
import { normalizeError } from '@/lib/errors/error-handler';
import { Notification, NotificationType, NotificationMetadata } from '@ramyas-jeweller/shared-types';

export interface CreateNotificationPayload {
  customerId?: string;
  title: string;
  message: string;
  type: NotificationType;
  metadata?: NotificationMetadata;
}

export class NotificationService {
  private static getSupabase() {
    return createClient();
  }

  /**
   * Helper to map database notification row to Notification interface
   */
  private static mapRowToNotification(row: Record<string, unknown>): Notification {
    return {
      id: String(row.id),
      customerId: row.customer_id ? String(row.customer_id) : null,
      title: String(row.title ?? ''),
      message: String(row.message ?? ''),
      type: (row.type as NotificationType) ?? 'ANNOUNCEMENT',
      isRead: Boolean(row.is_read ?? false),
      metadata: (row.metadata as NotificationMetadata) ?? null,
      createdAt: String(row.created_at ?? new Date().toISOString()),
    };
  }

  /**
   * Fetch recent notifications from public.notifications
   */
  static async getNotifications(limit: number = 20): Promise<Notification[]> {
    const supabase = this.getSupabase();

    try {
      const { data, error } = await supabase
        .from('notifications')
        .select('*')
        .order('created_at', { ascending: false })
        .limit(limit);

      if (error) {
        throw normalizeError(error);
      }

      return (data ?? []).map((row) => this.mapRowToNotification(row));
    } catch (error) {
      throw normalizeError(error);
    }
  }

  /**
   * Get unread notification count
   */
  static async getUnreadNotificationCount(): Promise<number> {
    const supabase = this.getSupabase();

    try {
      const { count, error } = await supabase
        .from('notifications')
        .select('*', { count: 'exact', head: true })
        .eq('is_read', false);

      if (error) {
        throw normalizeError(error);
      }

      return count ?? 0;
    } catch (error) {
      throw normalizeError(error);
    }
  }

  /**
   * Mark a single notification as read
   */
  static async markNotificationAsRead(id: string): Promise<Notification> {
    const supabase = this.getSupabase();

    try {
      if (!id) {
        throw new AppError('Notification ID is required.', ErrorCode.VALIDATION_ERROR, 400);
      }

      const { data, error } = await supabase
        .from('notifications')
        .update({ is_read: true })
        .eq('id', id)
        .select('*')
        .single();

      if (error) {
        throw normalizeError(error);
      }

      return this.mapRowToNotification(data);
    } catch (error) {
      throw normalizeError(error);
    }
  }

  /**
   * Mark all unread notifications as read
   */
  static async markAllNotificationsAsRead(): Promise<boolean> {
    const supabase = this.getSupabase();

    try {
      const { error } = await supabase
        .from('notifications')
        .update({ is_read: true })
        .eq('is_read', false);

      if (error) {
        throw normalizeError(error);
      }

      return true;
    } catch (error) {
      throw normalizeError(error);
    }
  }

  /**
   * Create a new in-app system notification
   */
  static async createNotification(payload: CreateNotificationPayload): Promise<Notification> {
    const supabase = this.getSupabase();

    try {
      if (!payload.title || !payload.message) {
        throw new AppError('Title and message are required for notification.', ErrorCode.VALIDATION_ERROR, 400);
      }

      const { data, error } = await supabase
        .from('notifications')
        .insert({
          customer_id: payload.customerId || null,
          title: payload.title.trim(),
          message: payload.message.trim(),
          type: payload.type || 'ANNOUNCEMENT',
          metadata: payload.metadata || {},
          is_read: false,
        })
        .select('*')
        .single();

      if (error) {
        const msg = error.message;
        if (msg.includes('row-level security') || msg.includes('violates row-level security')) {
          throw new AppError('Access denied. Owner, Admin, or Staff role required to create notifications.', ErrorCode.FORBIDDEN, 403);
        }
        throw normalizeError(error);
      }

      return this.mapRowToNotification(data);
    } catch (error) {
      throw normalizeError(error);
    }
  }
}

File: apps/customer-app/lib/features/notifications/services/notification_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

class CustomerNotificationModel {
  final String id;
  final String? customerId;
  final String title;
  final String message;
  final String type;
  final bool isRead;
  final Map<String, dynamic>? metadata;
  final DateTime createdAt;

  CustomerNotificationModel({
    required this.id,
    this.customerId,
    required this.title,
    required this.message,
    required this.type,
    required this.isRead,
    this.metadata,
    required this.createdAt,
  });

  factory CustomerNotificationModel.fromJson(Map<String, dynamic> json) {
    return CustomerNotificationModel(
      id: json['id'] as String,
      customerId: json['customer_id'] as String?,
      title: json['title'] as String? ?? 'Notification',
      message: json['message'] as String? ?? '',
      type: json['type'] as String? ?? 'ANNOUNCEMENT',
      isRead: json['is_read'] as bool? ?? false,
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }
}

class CustomerNotificationService {
  static SupabaseClient get _supabase => Supabase.instance.client;

  /// Fetch notifications from public.notifications for current customer
  static Future<List<CustomerNotificationModel>> getNotifications() async {
    try {
      final response = await _supabase
          .from('notifications')
          .select()
          .order('created_at', ascending: false);

      final List<dynamic> data = response as List<dynamic>;
      return data.map((json) => CustomerNotificationModel.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      return [];
    }
  }

  /// Get unread notification count
  static Future<int> getUnreadCount() async {
    try {
      final response = await _supabase
          .from('notifications')
          .select('id')
          .eq('is_read', false);

      final List<dynamic> data = response as List<dynamic>;
      return data.length;
    } catch (e) {
      return 0;
    }
  }

  /// Mark single notification as read
  static Future<void> markAsRead(String id) async {
    try {
      await _supabase
          .from('notifications')
          .update({'is_read': true})
          .eq('id', id);
    } catch (e) {
      // Handle error cleanly
    }
  }

  /// Mark all notifications as read
  static Future<void> markAllAsRead() async {
    try {
      await _supabase
          .from('notifications')
          .update({'is_read': true})
          .eq('is_read', false);
    } catch (e) {
      // Handle error cleanly
    }
  }
}
